### PR TITLE
issue-3467 Add option to disable Network Policy Generation.

### PR DESCRIPTION
**Pull request by davidjlynn**
### Type of change

- Enhancement / new feature

### Description

This pull request aims to solve #3467 by adding a feature gate to toggle the generation of network policy generation by Strimzi. This feature gate is by default enabled (as this maintains current behaviour) and can be used by specifying the `STRIMZI_NETWORK_POLICY_GENERATION` environment variable as `false` in the configuration.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Comment by davidjlynn** [comment]
As mentioned in strimzi/proposals#35 an environment variable should be used instead.

This PR switched over to this approach, rather than the previous approach of feature gate.

File: documentation/modules/operators/ref-operator-cluster.adoc [hunk 1]
```diff
@@ -168,6 +168,14 @@ If you are using a different DNS domain name suffix in your cluster, change the
 The timeout for building new Kafka Connect images with additional connectots, in milliseconds.
 This value should be increased when using Strimzi to build container images containing many connectors or using a slow container registry.
 
+`STRIMZI_NETWORK_POLICY_GENERATION` :: Optional, default `true`.
+The boolean value, controlling whether Strimzi should generate Network Policy files using the operator.
```
Comment (scholzj) [inline documentation/modules/operators/ref-operator-cluster.adoc:172]: Maybe `Network Policy resources` is better than `Network Policy files`?

**Review by scholzj — COMMENTED** [review]
The way it is implemented here is that it will always reconcile the network policies. This has two side-effects:
* If your custom network policy has the same name as the one generated by the operator, the operator will wipe it.
* It still requires the RBAC rights to work with network policies. That means it will not work in use-cases where the user is not allowed to use network policies (we had such as well).

I wonder if it would be better to instead of changing the models, to instead not reconcile the network policies. Either by using some _dummy_ NetworkPolicyOperator would would do nothing (that might make it easier to unit test). Or by handling it in the operator code for example like this (from the `KafkaAssemblyOperator` class):

```
        Future<ReconciliationState> kafkaNetPolicy() {
            if (isNetworkPolicyGeneration) {
                return withVoid(networkPolicyOperator.reconcile(reconciliation, namespace, KafkaCluster.networkPolicyName(name), kafkaCluster.generateNetworkPolicy(operatorNamespace, operatorNamespaceLabels)));
            } else {
                return withVoid(Future.succeededFuture());
            }
        }
```

**Comment by davidjlynn** [comment]
> I wonder if it would be better to instead of changing the models, to instead not reconcile the network policies. Either by using some _dummy_ NetworkPolicyOperator would would do nothing (that might make it easier to unit test). Or by handling it in the operator code for example like this (from the `KafkaAssemblyOperator` class)

@scholzj I am happy to switch to the example you gave above, indeed this avoids the reconciliation step completely which is sensible.

File: documentation/modules/operators/ref-operator-cluster.adoc (hunk 1 above)
Comment (davidjlynn) [reply documentation/modules/operators/ref-operator-cluster.adoc]: Switched to better naming.

**Review by scholzj — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,7 @@
 * Open Policy Agent authorizer updated to a new version supporting Scala 2.13. See the _Changes, deprecations and removals_ sections for more details. (#5192)
 * Allow a custom password to be set for SCRAM-SHA-512 users by referencing a secret in the `KafkaUser` resource
 * Add support for `tls-external` authentication to User Operator to allow management of ACLs and Quotas for TLS users with user certificates generated externally (#5249) 
+* Add support for disabling generation of network policies (Available via the `STRIMZI_NETWORK_POLICY_GENERATION` environment variable, disabling this will disable network policies)
```
Comment (PaulRMellor) [inline CHANGELOG.md:9]:
```suggestion
* Support for disabling the automatic generation of network policies for listeners. Users can now use the Cluster Operator's `STRIMZI_NETWORK_POLICY_GENERATION` environment variable to disable network policies.
```

Comment (davidjlynn) [reply CHANGELOG.md:9]: Thanks @PaulRMellor , I like your phrasing better but just wanted to also remove the "for listeners" part.
This applies to all Network Policies and the Cluster Operator generates policies for Metrics too.
I have change to:
`* Support for disabling the automatic generation of network policies by the Cluster Operator. Set the Cluster Operator's `STRIMZI_NETWORK_POLICY_GENERATION` environment variable to `false` to disable network policies.`
Let me know if this is acceptable.

File: documentation/modules/operators/ref-operator-cluster.adoc [hunk 3]
```diff
@@ -168,6 +168,14 @@ If you are using a different DNS domain name suffix in your cluster, change the
...
 This value should be increased when using Strimzi to build container images containing many connectors or using a slow container registry.
 
+`STRIMZI_NETWORK_POLICY_GENERATION` :: Optional, default `true`.
+The boolean value, controlling whether Strimzi should generate Network Policy resources using the operator.
+
+By default, this is enabled, causing Network policies to be generated which allow Kafka connections between the relevant components involved.
+
+However, if the user desires custom network policies, they may disable this generation and take the responsibility for maintaining connections between the components.
+
+
```
Comment (PaulRMellor) [inline documentation/modules/operators/ref-operator-cluster.adoc:178]:
```suggestion
`STRIMZI_NETWORK_POLICY_GENERATION` :: Optional, default `true`.
Controls whether Strimzi generates network policy resources.
Network policies allow connections between Kafka components.

Set this environment variable to `false` to disable network policy generation. You might do this, for example, if you want to use custom network policies. Custom network policies allow more control over maintaining the connections between components.

```

**Review by PaulRMellor — APPROVED** [review]
Looks good. Couple of suggestions

File: CHANGELOG.md (hunk 2 above)
Comment (PaulRMellor) [reply CHANGELOG.md]: Thanks @davidjlynn - i took out the listener part when I realised what this was about. :-) See above^

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,7 @@
 * Open Policy Agent authorizer updated to a new version supporting Scala 2.13. See the _Changes, deprecations and removals_ sections for more details. (#5192)
 * Allow a custom password to be set for SCRAM-SHA-512 users by referencing a secret in the `KafkaUser` resource
 * Add support for `tls-external` authentication to User Operator to allow management of ACLs and Quotas for TLS users with user certificates generated externally (#5249) 
+* Support for disabling the automatic generation of network policies by the Cluster Operator. Set the Cluster Operator's `STRIMZI_NETWORK_POLICY_GENERATION` environment variable to `false` to disable network policies.
```
Comment (PaulRMellor) [inline CHANGELOG.md:9]:
```suggestion
* Support for disabling the automatic generation of network policies by the Cluster Operator. Set the Cluster Operator's `STRIMZI_NETWORK_POLICY_GENERATION` environment variable to `false` to disable network policies. (#5258)
```

**Comment by scholzj** [comment]
@davidjlynn Can you have a look at the broken DCO sign-off? Under the _Details_ link, there should be some instructions how to fix it.

**Review by tombentley — APPROVED** [review]
Thanks!

**Comment by davidjlynn** [comment]
I have messed up the squash, (The sign off command wasn't working) I will fix ASAP.

**Comment by davidjlynn** [comment]
Squash and DCO should be fixed now.

**Comment by scholzj** [comment]
Thanks. I will merge it after the tests pass.

**Pull request merged**